File: android/src/main/kotlin/com/sri/amount_localizer/AmountLocalizerPlugin.kt
package com.sri.amount_localizer

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.Locale

/**
 * Exposes the device's FORMAT locale country — the setting that actually
 * drives Gboard's decimal key — to Flutter.
 *
 * Flutter's PlatformDispatcher only reports the *language* locale. On a
 * device set to English language with Brazil format region, Flutter sees
 * `en_US` while the numeric keyboard emits `,`. Reading
 * `Locale.getDefault(Locale.Category.FORMAT).country` closes that gap.
 */
class AmountLocalizerPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == METHOD_GET_FORMAT_REGION) {
            val country = Locale.getDefault(Locale.Category.FORMAT).country
            result.success(if (country.isNullOrEmpty()) null else country)
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private companion object {
        const val CHANNEL_NAME = "com.sri.amount_localizer/region"
        const val METHOD_GET_FORMAT_REGION = "getFormatRegion"
    }
}

File: ios/Classes/AmountLocalizerPlugin.swift
import Flutter
import UIKit

/// Exposes the iOS Region setting — the setting that actually drives the
/// numeric-keyboard decimal key — to Flutter.
///
/// Flutter's `PlatformDispatcher` only reports the *language* locale. On a
/// device set to English language with Brazil region, Flutter sees `en_US`
/// while the numeric keyboard emits `,`. Reading `Locale.current.region`
/// (iOS 16+) / `Locale.current.regionCode` (older) closes that gap.
public class AmountLocalizerPlugin: NSObject, FlutterPlugin {
  private static let channelName = "com.sri.amount_localizer/region"
  private static let methodGetFormatRegion = "getFormatRegion"

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )
    let instance = AmountLocalizerPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if call.method == AmountLocalizerPlugin.methodGetFormatRegion {
      let code: String?
      if #available(iOS 16.0, *) {
        code = Locale.current.region?.identifier
      } else {
        code = Locale.current.regionCode
      }
      result(code)
    } else {
      result(FlutterMethodNotImplemented)
    }
  }
}
